File: frontend/types/Patrimony.ts
// types/Patrimony.ts
// types/Patrimony.ts

export interface PatrimonyItem {
  id: string;
  plate: string;
  name: string;
  description: string;
  acquisitionDate: string;
  value: number;
  department: string;
  status: 'active' | 'inactive' | 'maintenance' | 'written_off';
  imageUrl?: string;
  sector?: string;
  nfIssueDate?: string;
  supplier?: string;
  invoiceNumber?: string;
  commitmentNumber?: string;
  denfSeNumber?: string;
  invoiceFile?: string;
  commitmentFile?: string;
  denfSeFile?: string;
  createdBy?: string;
  createdAt: string;
  updatedAt: string;
  isVehicle?: boolean;
  fiscalDocuments?: FiscalDocument[];
  // ✅ ADICIONE OS NOVOS CAMPOS
  fleetNumber?: string;
  fleetNotes?: string;
}

// =========================
// DOCUMENTOS FISCAIS
// =========================
export interface FiscalDocument {
  id?: string;
  invoiceNumber: string;
  invoiceFile?: string;
  invoiceFileName?: string;
  commitmentNumber: string;
  commitmentFile?: string;
  commitmentFileName?: string;
  issueDate?: string;
  isLegacy?: boolean;

  _invoiceFile?: File;
  _commitmentFile?: File;
}

// =========================
// TRANSFERÊNCIA
// =========================
export interface TransferRequest {
  patrimonyId: string;
  fromDepartment: string;
  toDepartment: string;
  reason: string;
}

// =========================
// DEPARTAMENTO
// =========================
export interface Department {
  id: string;
  name: string;
  manager: string;
  phone: string;
}

export interface DepartmentStats {
  department: string;
  count: number;
  totalValue: number;
}

export interface Stats {
  total: number;
  active: number;
  inactive: number;
  maintenance: number;
  writtenOff: number;
  totalValue: number;
  byDepartment: DepartmentStats[];
}

// =========================
// USUÁRIO
// =========================
export interface User {
  id: string;
  username: string;
  email?: string;
  role: string;
  company_name: string;
  department: string;
  created_at: string;
  updated_at: string;
}

export interface LoginResponse {
  token: string;
  user: User;
}

export interface CreateUser {
  company_name: string;
  department: string;
  username: string;
  password: string;
  email?: string;
  role?: string;
}

export interface LoginRequest {
  username: string;
  password: string;
}

// =========================
// FORMULÁRIO
// =========================
export interface PatrimonyFormData {
  plate: string;
  name: string;
  description: string;
  acquisition_date: string;
  value: string;
  department: string;
  status: string;
  sector?: string;
  nf_issue_date?: string;
  supplier?: string;
  invoice_number?: string;
  commitment_number?: string;
  denf_se_number?: string;
  is_vehicle?: boolean;
}

// =========================
// 🚨 FROTA CORRIGIDA (IMPORTANTE)
// =========================
export interface FleetItem {
  id: string;

  fleet_number: string;
  patrimony_id: string;

  // 👇 nomes corretos usados no print
  patrimony_plate?: string;
  patrimony_name?: string;
  patrimony_description?: string;

  department: string;

  // 👇 AGORA EXISTE PARA NÃO QUEBRAR SEU PRINT
  sector?: string;

  notes?: string;

  created_at: string;
  updated_at: string;
}

// =========================
// FORM FROTA
// =========================
export interface FleetFormData {
  fleetNumber: string;
  patrimonyId: string;
  department: string;
  notes: string;
}

// =========================
// ERRO API
// =========================
export interface ApiError {
  message: string;
  field?: string;
  type?: string;
}

// =========================
// FILTROS
// =========================
export interface SearchFilters {
  plate: string;
  name: string;
  department: string;
  status: string;
  sector: string;
  supplier: string;
}

File: backend/src/models/patrimony.rs
//backend/src/models/patrimony.rs
use serde::{Deserialize, Serialize};
use chrono::{Utc, NaiveDate};
use uuid::Uuid;
// Adicione após a definição existente de UpdatePatrimony

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FiscalDocument {
    pub id: Uuid,
    pub patrimony_id: Uuid,
    pub invoice_number: Option<String>,
    pub commitment_number: Option<String>,
    pub invoice_file: Option<String>,
    pub commitment_file: Option<String>,
    pub nf_issue_date: Option<NaiveDate>,
    pub supplier: Option<String>,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct FiscalDocumentRequest {
    pub invoice_number: Option<String>,
    pub commitment_number: Option<String>,
    pub invoice_file: Option<String>,
    pub commitment_file: Option<String>,
    pub nf_issue_date: Option<NaiveDate>,
    pub supplier: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FleetItem {
    pub id: Uuid,
    pub fleet_number: String,
    pub patrimony_id: Option<Uuid>,
    pub patrimony_plate: Option<String>,
    pub patrimony_name: Option<String>,
    pub department: String,
    pub notes: Option<String>,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct CreateFleet {
    pub fleet_number: String,
    pub patrimony_id: Option<Uuid>,
    pub department: String,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateFleet {
    pub fleet_number: Option<String>,
    pub patrimony_id: Option<Uuid>,
    pub department: Option<String>,
    pub notes: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Patrimony {
    pub id: Uuid,
    pub plate: String,
    pub name: String,
    pub description: String,
    pub acquisition_date: NaiveDate,
    pub value: f64,
    pub department: String,
    pub status: String,
    pub invoice_number: Option<String>,          // ✅ Campo de documento
    pub commitment_number: Option<String>,       // ✅ Campo de documento
    pub denf_se_number: Option<String>,          // ✅ Campo de documento
    pub invoice_file: Option<String>,            // ✅ Arquivo NF
    pub commitment_file: Option<String>,         // ✅ Arquivo empenho
    pub denf_se_file: Option<String>,            // ✅ Arquivo DENF/SE
    pub image_url: Option<String>,
    pub sector: Option<String>,                  // ✅ NOVO CAMPO: Setor
    pub nf_issue_date: Option<NaiveDate>,        // ✅ NOVO CAMPO: Data emissão NF
    pub supplier: Option<String>,                // ✅ NOVO CAMPO: Fornecedor
    pub is_vehicle: bool,                        // ✅ NOVO CAMPO: flag para veículos
    pub created_by: Option<Uuid>,                // user who created the entry
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
    // ⚠️ ADICIONE ESTE CAMPO:
    pub fiscal_documents: Option<Vec<FiscalDocument>>,
}

#[derive(Deserialize)]
pub struct CreatePatrimony {
    pub plate: String,
    pub name: String,
    pub description: String,
    pub acquisition_date: NaiveDate,
    pub value: f64,
    pub department: String,
    pub status: String,
    pub invoice_number: Option<String>,          // ✅ Campo de documento
    pub commitment_number: Option<String>,       // ✅ Campo de documento
    pub denf_se_number: Option<String>,          // ✅ Campo de documento
    pub sector: Option<String>,                  // ✅ NOVO CAMPO: Setor
    pub nf_issue_date: Option<NaiveDate>,        // ✅ NOVO CAMPO: Data emissão NF
    pub supplier: Option<String>,                // ✅ NOVO CAMPO: Fornecedor
    pub is_vehicle: Option<bool>,                 // indica se o bem é veículo
    // ⚠️ ADICIONE TAMBÉM ESTE CAMPO no CreatePatrimony:
    pub fiscal_documents: Option<Vec<FiscalDocumentRequest>>,
}

#[derive(Deserialize)]
pub struct UpdatePatrimony {
    pub plate: Option<String>,
    pub name: Option<String>,
    pub description: Option<String>,
    pub acquisition_date: Option<NaiveDate>,
    pub value: Option<f64>,
    pub department: Option<String>,
    pub status: Option<String>,
    pub invoice_number: Option<String>,          // ✅ Campo de documento
    pub commitment_number: Option<String>,       // ✅ Campo de documento
    pub denf_se_number: Option<String>,          // ✅ Campo de documento
    pub sector: Option<String>,                  // ✅ NOVO CAMPO: Setor
    pub nf_issue_date: Option<NaiveDate>,        // ✅ NOVO CAMPO: Data emissão NF
    pub supplier: Option<String>,                // ✅ NOVO CAMPO: Fornecedor
    pub is_vehicle: Option<bool>,                 // flag para veículos
     // ⚠️ ADICIONE TAMBÉM ESTE CAMPO no UpdatePatrimony:
    pub fiscal_documents: Option<Vec<FiscalDocumentRequest>>,   
}
